run/pem2john.py

```diff
@@ -108,7 +103,10 @@ def unwrap_pkcs8_data(blob):
         encrypted_datah = hexlify(encrypted_data).decode("ascii")
         ivh = hexlify(iv).decode("ascii")
 
-        sys.stdout.write("$PEM$1$%d$%s$%s$%s$%d$%s\n" % (cid, salth, iterations, ivh, len(encrypted_data), encrypted_datah))
+        pem_version = "$PEM$1"
+        if kdf_params["prf"]["algorithm"] != "sha1":
+            pem_version = "$PEM$2$sha256"
```

Comment:
Here it is assumed that if it is not sha1 then it must be sha256. I doubt that this assumption always holds. What does `kdf_params["prf"]["algorithm"]` contain in your test? Is it literally `sha256`? Then this value, whatever it would be, could be part of the format line. This would make the tool more future-proof.

However, this also requires to test within the C code which algorithm is specified. And then reject all but the supported hashes.

Reply:
IMHO, this PR only checks the type in `john` proper, so adding sha256 is redundant. Just using `pem_version = "$PEM$2" ` for anything that is not sha1 will suffice.

Reply:
I'm with @exploide on this. The new format should be future-proof, so we wouldn't need another change to the script if a new `prf` type is used later. Also, claiming `sha256` for anything other than `sha1` would even lead to false negatives under that scenario - so it is a bug.

In fact, we may want to choose to encode the KDF type as well, and maybe encode the cipher type as-is without converting it to `cid` (oh, and by the way it's a bug we didn't change the `return False` to `return True` for the cipher in the previous PR). So we'd have strings like `pbkdf2-sha256-aes256_cbc`, which we'd need to parse in `john`.

Reply:
Okay, let me see if I understand. Currently our string formatting looks like `$PEM$1$CID` and your suggestion is changing that to `$PEM$prf$cipher-type`, so the strings will look like `$PEM$sha1$aes256_cdc `or `$PEM$sha256$aes256_cdc`. Obviously this will result in a lot of changes in the way we parse in `john`. The behaviour in `john` will be as-is for strings like `$PEM$sha1$aes256_cdc` but it should refuse strings like `$PEM$sha256$aes256_cdc`. Did I get that right? :thinking:

Reply:
Close, but not exactly. I suggest that for compatibility of new `pem2john.py` with old `john`, we continue to output the `$PEM$1` encoding for `sha1`. When the combination of KDF/prf/cipher does not fit in the old encoding (which currently means `prf` is `sha256`, but in the future can possibly also happen in other cases), output `$PEM$2$kdf$prf$cipher`.

Initially, the "format" in `john` would reject anything other than `$PEM$1`, so you don't strictly have to parse the extra fields right away (in this PR) - although you can in order to produce the same kinds of different messages that the script used to produce.

In your next PR, you'd actually implement support for `prf` of `sha256`.